external/mbedtls/alt/pk_wrap_alt.c

```diff
@@ -827,88 +888,145 @@ int eckey_verify_wrap(void *ctx, mbedtls_md_type_t md_alg, const unsigned char *
+	}
+	curve = ecdsa.grp.id;
+	mbedtls_ecdsa_free(&ecdsa);
 
-		hal_data t_hash = {0,};
-		t_hash.data = (unsigned char *)malloc(hash_len);
-		if (t_hash.data == NULL) {
-			return MBEDTLS_ERR_ECP_ALLOC_FAILED;
-		}
-		memcpy(t_hash.data, hash, hash_len);
-		t_hash.data_len = hash_len;
+	/*
+	 * 1. Check hash algorithm and sign curve
+	 */
+	if ((ret = _eckey_get_ecdsa_mode(md_alg, curve, &ecdsa_mode)) != 0) {
+		return ret;
+	}
 
-		mbedtls_ecdsa_free(&ecdsa);
-		/*
-		 * 1. Check hash algorithm and sign curve
-		 */
-		ret = ecdsa_get_mode(md_alg, curve, &ecdsa_mode);
-		if (ret) {
-			free(t_hash.data);
-			return ret;
-		}
+	if ((key_type = alt_get_keytype(curve)) == HAL_KEY_UNKNOWN) {
+		return MBEDTLS_ERR_ECP_HW_ACCEL_FAILED;
+	}
 
-		/*
-		 * 2. Sign
-		 */
-		hal_data sign = {sig, 0, NULL, 0};
+	/* If key index is not in injected key range  0 ~ 31
+	 * then key is set by mbedtls_pk_parse_key() */
+	if ((((mbedtls_ecp_keypair *)ctx)->key_index) == 0xff) {
+		is_injected_key = 0;
+	}
 
-		ret = sl_init(&shnd);
-		if (ret != SECLINK_OK) {
-			free(t_hash.data);
+	if (!is_injected_key) {
+		/* application doesn't set injected key.
+		 * So the private key are inserted by buffer. */
+		prikey_buf = (unsigned char *)malloc(pubkey_buflen);
+		if (prikey_buf == NULL) {
+			return MBEDTLS_ERR_ECP_HW_ACCEL_FAILED;
+		}
+		/* Convert a private key to DER*/
+		int len = _eckey_write_key_der((mbedtls_ecp_keypair *)ctx, prikey_buf, pubkey_buflen);
+		if (len <= 0) {
+			free(prikey_buf);
+			return MBEDTLS_ERR_ECP_INVALID_KEY;
+		}
+		prikey.data_len = len;
+		prikey.data = (unsigned char *)malloc(len);
+		if (!prikey.data) {
+			free(prikey_buf);
 			return MBEDTLS_ERR_ECP_HW_ACCEL_FAILED;
 		}
+		memcpy(prikey.data, prikey_buf + pubkey_buflen - len, len);
+		free(prikey_buf);
+	}
+
+	t_hash.data = (unsigned char *)malloc(hash_len);
+	if (t_hash.data == NULL) {
+		if (!is_injected_key) {
+			free(prikey.data);
+		}
+		return MBEDTLS_ERR_ECP_ALLOC_FAILED;
+	}
+	memcpy(t_hash.data, hash, hash_len);
+	t_hash.data_len = hash_len;
 
-		ret = sl_ecdsa_sign_md(shnd, ecdsa_mode, &t_hash, key_idx, &sign, &hres);
+	/*
+	 * 2. Sign
+	 */
+	if ((ret = sl_init(&shnd)) != SECLINK_OK) {
+		if (!is_injected_key) {
+			free(prikey.data);
+		}
 		free(t_hash.data);
-		if (ret != SECLINK_OK || hres != HAL_SUCCESS) {
+		return MBEDTLS_ERR_ECP_HW_ACCEL_FAILED;
+	}
+
+	if (!is_injected_key) {
+		key_idx = alt_set_key(shnd, key_type, NULL, &prikey, 32);
+		free(prikey.data);
+		if (key_idx < 0) {
+			free(t_hash.data);
 			sl_deinit(shnd);
 			return MBEDTLS_ERR_ECP_HW_ACCEL_FAILED;
 		}
+	}
 
-		/* Use r and s to generate signature */
-		*sig_len = sign.data_len;
-		sl_deinit(shnd);
-		return ret;
+	ret = sl_ecdsa_sign_md(shnd, ecdsa_mode, &t_hash, key_idx, &sign);
+	if (!is_injected_key) {
+		sl_remove_key(shnd, key_type, key_idx);
+	}
+	sl_deinit(shnd);
+	free(t_hash.data);
+
+	if (ret != SECLINK_OK) {
+		return MBEDTLS_ERR_ECP_HW_ACCEL_FAILED;
```

Comment:
Is it okay not to free the `prikey.data`?

Reply:
pirkey is freed at 1008 line